### PR TITLE
Support identifier-first unlock account flow

### PR DESCRIPTION
**Pull request by trevoring-okta**
## Description:
**Context:** AuthExp team released backend support for identifier-first unlock account flow. Gen2 is able to support the updates to IDX more or less out of the box, but Gen3 already includes a faked version of identifier-first unlock account using Stepper-based logic (original PR was #3414)
- Adds transformers/views for the `unlock-account` remediation type, which was previously unused but now maps to the identifier-only view within the flow
- Forks the gen3 `transformSelectAuthenticatorUnlockVerify` transformer to use the updated logic if the IDX payload does not have an `identifier` field object or the previous remediation name was `unlock-account`. Otherwise, it continues to use the faked logic
- Converts gen3-only unlock account spec to instead be an identifier-first spec since identifier-first is no longer only present on gen3

## PR Checklist

- [ ] Have you verified the basic functionality for this change?
- [ ] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-657627](https://oktainc.atlassian.net/browse/OKTA-657627)

#### Identifier-first flow in gen2
https://github.com/okta/okta-signin-widget/assets/106110543/a7315c55-eb44-47ad-8373-b567269af71f

#### Identifier-first flow in gen3
https://github.com/okta/okta-signin-widget/assets/106110543/91359795-8a6a-4189-ad45-d85efd84f987

### CODE REVIEW

File: src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorUnlockVerify.ts [hunk 1]
```diff
@@ -78,8 +67,36 @@ export const transformSelectAuthenticatorUnlockVerify: IdxStepTransformer = ({
...
   const identifier = getUIElementWithName('identifier', uischema.elements) as FieldElement;
 
+  // TODO: OKTA-741832 Remove this fork once AUTH_POLICY_FOR_ACCOUNT_MANAGEMENT monolith FF goes GA
+  // Identifier-first unlock account flow was initially 'faked' in this transformer and released
+  // w/o backend changes. The backend changes ensure that this transformer is never hit with
+  // an identifier UI schema element in the IDX response, so this can be used as a logical fork
+  if (!identifier || prevStepName === IDX_STEP.UNLOCK_ACCOUNT) {
+    uischema.elements.unshift(verifyTitle, verifySubtitle, authenticatorList);
+    return formBag;
+  }
```
Comment (trevoring-okta) [inline src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorUnlockVerify.ts:84]: Logical fork can be cleaned up once `AUTH_POLICY_FOR_ACCOUNT_MANAGEMENT` FF goes GA

File: src/v3/jest.config.js [hunk 2]
```diff
@@ -22,7 +22,7 @@ const esModules = [
   '@okta/odyssey-react-mui',
   '@adobe/leonardo-contrast-colors',
   'apca-w3',
-  'colorparsley'
+  'colorparsley',
```
Comment (trevoring-okta) [inline src/v3/jest.config.js:25]: Unrelated lint fix that I pulled into this PR, can disregard for review purposes

File: src/v3/src/transformer/oktaVerify/transformOktaVerifyEnrollPoll.ts [hunk 3]
```diff
@@ -439,7 +439,7 @@ export const transformOktaVerifyEnrollPoll: IdxStepTransformer = ({
 
   const stepper: StepperLayout = {
     type: UISchemaLayoutType.STEPPER,
-    key: 'stepper_' + channelType,
+    key: `stepper_${channelType}`,
```
Comment (trevoring-okta) [inline src/v3/src/transformer/oktaVerify/transformOktaVerifyEnrollPoll.ts:442]: Another unrelated lint fix

**Comment by denysoblohin-okta** [comment]
Do changes to Gen2 break monolith tests?

File: src/v2/ion/i18nUtils.js [hunk 4]
```diff
@@ -30,6 +30,7 @@ const SECURITY_QUESTION_PREFIXES = [
 const I18N_OVERRIDE_MAPPINGS = {
   'identify.identifier': 'primaryauth.username.placeholder',
   'select-authenticator-unlock-account.identifier': 'primaryauth.username.placeholder',
+  'unlock-account.identifier': 'primaryauth.username.placeholder',
```
Comment (justinabrokwah-okta) [inline src/v2/ion/i18nUtils.js:33]: Which string does this override?

**Comment by trevoring-okta** [comment]
> Do changes to Gen2 break monolith tests?

@denysoblohin-okta Downstream came back green: https://bacon-go.aue1e.saasure.net/commits?artifact=monolith&branch=d16t-okta-signin-widget-0355903-667ace9f&page=1&pageSize=6&sha=5f7b6331ed1e6f4994a44662a2a23f782c8bae8a&tab=main

File: test/testcafe/spec/UnlockAccountChallengeIdentifierFirst_spec.js [hunk 5]
```diff
@@ -124,13 +173,45 @@ test.requestHooks(identifyLockedUserMock)('should render custom Unlock account l
...
+
+
 test.requestHooks(identifyLockedUserMock)('should show unlock account authenticator selection list when user has more than one authenticator', async t => {
   const identityPage = await setup(t);
   await checkA11y(t);
   await identityPage.clickUnlockAccountLink();
 
   const selectFactorPage = new SelectFactorPageObject(t);
-  await t.expect(selectFactorPage.getFormTitle()).eql('Unlock account?');
+  await t.expect(selectFactorPage.getFormTitle()).contains('Unlock account');
```
Comment (justinabrokwah-okta) [inline test/testcafe/spec/UnlockAccountChallengeIdentifierFirst_spec.js:214]: Why did this change to `contains`?

**Review by justinabrokwah-okta — COMMENTED** [review]
Looks good overall, just left a few questions

File: test/testcafe/spec/UnlockAccountChallenge_spec.js [hunk 6]
```diff
@@ -153,7 +152,7 @@ test.requestHooks(errorUnlockAccount)('should show error if identifier is blank'
   await checkA11y(t);
   await identityPage.clickUnlockAccountLink();
   const selectFactorPage = new SelectFactorPageObject(t);
-  await t.expect(selectFactorPage.getFormTitle()).eql('Unlock account?');
+  await t.expect(selectFactorPage.getFormTitle()).eql('Verify it\'s you with a security method');
```
Comment (shuowu-okta) [inline test/testcafe/spec/UnlockAccountChallenge_spec.js:155]: expected title change?

File: src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorUnlockVerify.ts [hunk 7]
```diff
@@ -78,8 +67,36 @@ export const transformSelectAuthenticatorUnlockVerify: IdxStepTransformer = ({
...
   };
 
+  const authenticatorList: AuthenticatorButtonListElement = {
+    type: 'AuthenticatorButtonList',
+    options: { buttons: authenticatorButtons, dataSe: 'authenticator-verify-list' },
+  };
+
   const identifier = getUIElementWithName('identifier', uischema.elements) as FieldElement;
 
+  // TODO: OKTA-741832 Remove this fork once AUTH_POLICY_FOR_ACCOUNT_MANAGEMENT monolith FF goes GA
```
[the remark this answers is not in the document]
Comment (trevoring-okta) [reply src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorUnlockVerify.ts:77]: It shouldn't have any time urgency behind it and can be made after GA. It's more of a tech debt item since once the FF goes GA we'll just end up with a dead code path for the faked flow

File: src/v2/ion/i18nUtils.js (hunk 4 above)
Comment (trevoring-okta) [reply src/v2/ion/i18nUtils.js:33]: This maps the `unlock-account` remediation name to the "Username" label above the identifier field so that label can be localized by our translation bundles. Before identifier-first support, we weren't using the `unlock-account` remediation so no mapping existed

File: playground/mocks/data/idp/idx/user-unlock-account-identifier-first.json [hunk 8]
```diff
@@ -0,0 +1,62 @@
+{
```
[the remark this answers is not in the document]
Comment (trevoring-okta) [reply playground/mocks/data/idp/idx/user-unlock-account-identifier-first.json:1]: Yeah you're right actually. I don't fully remember how it ended up this way but I think since the backend handles automatically challenging the 1 available authenticator if it sees that only one is available, the mock files end up being the same. I might have added this since they're used in different spec test flows, but tbh it's probably fine to just use the `user-unlock-account-identifier-first.json` mock for both 👍  I'll update that!

File: test/testcafe/spec/UnlockAccountChallengeIdentifierFirst_spec.js (hunk 5 above)
Comment (trevoring-okta) [reply test/testcafe/spec/UnlockAccountChallengeIdentifierFirst_spec.js:214]: Ah, so I changed those to support that gen2 uses the "Unlock account" string w/o a question mark but the faked gen3 flow was implemented with the "Unlock account?" string. That being said, it should actually be fine (and better) if I update the string for the faked gen3 flow to also use the string w/o a question mark. I'll update this as well, nice call out

File: test/testcafe/spec/UnlockAccountChallenge_spec.js (hunk 6 above)
Comment (trevoring-okta) [reply test/testcafe/spec/UnlockAccountChallenge_spec.js:155]: Yeah, the expected UX with the backend FF on is that this factor selection page should ideally render the authenticator button list with this "Verify it's you with a security method" string.

However, when the FF is off (which is what this spec test is simulating), we have decided that we're okay with living with this new string also being the title for the non-identifier-first flow.

This is what it looks like: 
Screenshot 2024-06-27 at 3 00 40 PM

File: src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorUnlockVerify.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (justinabrokwah-okta) [reply src/v3/src/transformer/selectAuthenticator/transformSelectAuthenticatorUnlockVerify.ts:77]: Ahh got it, TODO for tech debt not for GA; sounds good

**Review by justinabrokwah-okta — COMMENTED** [review]
LGTM so far, just waiting for a testing update from Ramya before approving

**Review by justinabrokwah-okta — APPROVED** [review]

**Review by shuowu-okta — APPROVED** [review]
